feat(Teacher): Implement unit tagging

geoffreykwan opened the pull request:
## Changes
- Added the ability to create, edit, and delete tags
- Added the ability to apply and remove tags from units

Run this mysql database query to add the color column to the user_tags table
```
alter table user_tags add column color varchar(25) after text;
```
## Test
- Test with https://github.com/WISE-Community/WISE-Client/pull/1813

Closes #259

hirokiterashima reviewed (changes requested):
The functionality mostly works. 

It looks like the number of tags in the message when you try to delete a tag is off. In the screenshot below, you can see that I have "tag3" on a unit, but the cofirmation message says "0" I think it may not be counting the tags that are on the units (as opposed to run units).

<img width="300" alt="Screenshot 2024-05-24 at 3 30 00 PM" src="https://github.com/WISE-Community/WISE-API/assets/119416/93d17a98-46d8-4ae6-aeb0-2bff1001f541">

I also added some code cleaning suggestions inline. 

Lastly, maybe we should also apply the code cleaning conventions for at least the new standalone components, like alphabetizing (imports -> selector -> standalone...), rename "styleUrls" -> "styleUrl", self-closing tags, and such?

hirokiterashima commented:
Oops. The comment above was intended for the WISE-Client PR. I have yet to review this PR.

hirokiterashima commented on src/main/java/org/wise/portal/presentation/web/controllers/tag/TagProjectController.java:
```diff
@@ -0,0 +1,88 @@
+    for (Project project : projects) {
+      userTagsService.applyTag(project, userTag);
+    }
+    return ResponseEntityGenerator.createSuccess(createProjectsResponse(user, projects));
```
Can this function just return a ```List...``` instead of ```ResponseEntity<List...>```?
```suggestion
    return createProjectsResponse(user, projects);
```

hirokiterashima commented on src/main/java/org/wise/portal/presentation/web/controllers/tag/TagProjectController.java:
```diff
@@ -0,0 +1,88 @@
+    for (Project project : projects) {
+      userTagsService.removeTag(project, userTag);
+    }
+    return ResponseEntityGenerator.createSuccess(createProjectsResponse(user, projects));
```
Can this function just return a ```List...``` instead of ```ResponseEntity<List...>```?
```suggestion
    return createProjectsResponse(user, projects);
```

hirokiterashima commented on src/main/java/org/wise/portal/presentation/web/controllers/tag/UserTagController.java:
```diff
@@ -0,0 +1,85 @@
+  private UserTagsService userTagsService;
+
+  @PostMapping("/user/tag")
+  protected ResponseEntity<Map<String, Object>> createTag(Authentication auth,
```
Instead of returning ResponseEntity, can we just return the Map object and throw a TagAlreadyExistsException if a tag exists?

hirokiterashima reviewed (commented):
Does this PR need to make changes to wise_db_init.sql?

Also, see inline comments. I wonder if the functions in [UserTag,TagProject]Controller should return the object directly instead of wrapping it in ```ResponseEntity```. I think Spring automatically returns 200 (unless there's an error), and 404 (if the function exists due to an exception)?

geoffreykwan commented:
I removed the ResponseEntity from TagProjectController.java. I kept it in the UserTagController.java because the client side looks for the error field to contain 'tagAlreadyExists' to show an error message. I've also updated wise_db_init.sql to include the color column in the user_tags table.

hirokiterashima commented:
For error handling- can we do something similar to how we handle invalid name when creating user accounts?
We throw an ```InvalidNameException``` in ```StudentAPIController.createStudentAccount()```, which is handled in register-student-form.component.createAccount(). It looks like we use ```@ControllerAdvice``` in CustomGlobalExceptionHandler.java to handle the InvalidNameException and return a ResponseEntity there. 

This will separate out the concerns and make the functions in UserTagController easier to read. They won't need to wrap the return value in ResponseEntityGenerator.createSuccess/createError each time.

geoffreykwan commented:
Ok, we now throw a TagAlreadyExistsException.

hirokiterashima commented:
See my inline comment in UserTagController. Do we need to use ResponseEntityGenerator?

geoffreykwan commented:
Ok, I've removed the usage of ResponseEntityGenerator.

hirokiterashima reviewed (approved):
LGTM

breity reviewed (approved):
Looks good. 👍

geoffreykwan commented:
:tada: This issue has been resolved in version 1.2.0 :tada:

The release is available on [GitHub release](https://github.com/WISE-Community/WISE-API/releases/tag/v1.2.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
